### PR TITLE
WT-6583 Introduce WT_TXN_BEGINNING to prevent the read timestamp from being cleared

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
With the change of 7b773c1, I found out that WT_TXN_RUNNING is only flagged for explicit transactions not automatic transactions. We need that check to free snapshot obtained by automatic transactions. So 7b773c1 doesn't fix the issue.

**Comment by quchenhao** [comment]
@agorrod @kommiharibabu @bvpvamsikrishna We haven't agreed on the solution for the problem yet. I want to gather feedback from you guys about this proposed solution. Personally, I prefer to fix the issue in `__wt_txn_begin` as I find I can't remove the check for WT_TXN_RUNNING flag in `__wt_txn_read_last` for the automatic transaction case. I introduce the new flag instead of flagging WT_TXN_RUNNING earlier at the start of `__wt_txn_begin` as I want to define the state of the transaction properly. Before we get the snapshot, we actually can't use the transaction so it is weird to mark it as running.

File: src/include/txn.h [hunk 1]
```diff
@@ -328,28 +328,29 @@ struct __wt_txn {
...
-#define WT_TXN_TS_COMMIT_ALWAYS 0x004000u
-#define WT_TXN_TS_COMMIT_KEYS 0x008000u
-#define WT_TXN_TS_COMMIT_NEVER 0x010000u
-#define WT_TXN_TS_DURABLE_ALWAYS 0x020000u
-#define WT_TXN_TS_DURABLE_KEYS 0x040000u
-#define WT_TXN_TS_DURABLE_NEVER 0x080000u
-#define WT_TXN_TS_ROUND_PREPARED 0x100000u
-#define WT_TXN_TS_ROUND_READ 0x200000u
-#define WT_TXN_UPDATE 0x400000u
+#define WT_TXN_BEGINNING 0x000002u /* Set only for explicit transactions. */
```
Comment (kommiharibabu) [inline src/include/txn.h:331]: How about `WT_TXN_STARTING`?

**Review by kommiharibabu — COMMENTED** [review]
lgtm. The new state of the transaction mentioning that is starting is a good way to represent while releasing the snapshot.

File: src/include/txn.h (hunk 1 above)
Comment (quchenhao) [reply src/include/txn.h:331]: The function name is `__wt_txn_begin`. To be consistent, I think beginning is better.

Comment (kommiharibabu) [reply src/include/txn.h:331]: Yes, agree to keep the flag same as with the function name.

**Comment by quchenhao** [comment]
@agorrod @bvpvamsikrishna What are your opinions?

**Comment by agorrod** [comment]
The proposed solution here doesn't feel like it will protect us from future similar bugs.

I think of the read timestamp as something a user has configured for the lifespan of a transaction. They don't expect that it will be cleared internally. WiredTiger calls `__wt_txn_release_snapshot` from eight different places, and that call will go on to unconditionally clear the read timestamp a user has set.

For example, WiredTiger will refresh (release and re-acquire) a snapshot for read-committed transactions. MongoDB could use a read-committed transaction for read only transactions (I'm not sure if they currently do, but they should be able to do that). With the code as structured we'd clear the read timestamp when refreshing the snapshot (I believe).

**Review by agorrod — CHANGES_REQUESTED** [review]
I've requested changes in a comment on the front page of the PR.

**Comment by kommiharibabu** [comment]
@agorrod @quchenhao How about we separate out the clearing the `read timestamp` as part of release snapshot and clear only when the transaction is either committed or rollbacked?

**Comment by quchenhao** [comment]
> @agorrod @quchenhao How about we separate out the clearing the `read timestamp` as part of release snapshot and clear only when the transaction is either committed or rollbacked? As per my understanding, the `read timestamp` is transaction property and have no relation to the snapshot.

Yes, I'm about to explore that.

**Comment by quchenhao** [comment]
Close in favor of #5999

**Pull request closed without merging**